### PR TITLE
Add renew token view

### PR DESCRIPTION
**Pull request by gberaudo**
If the same token as an existing one is recreated, reuse it.

### CODE REVIEW

File: c2corg_api/security/roles.py [hunk 1]
```diff
@@ -36,10 +36,14 @@ def validate_token(token):
         filter(Token.value == token and Token.expire > now).count() == 1
 
 
-def add_token(value, expire, userid):
-    token = Token(value=value, expire=expire, userid=userid)
-    DBSession.add(token)
-    DBSession.flush()
+def add_or_retrieve_token(value, expire, userid):
+    token = token = DBSession.query(Token). \
```
Comment (asaunier) [inline c2corg_api/security/roles.py]: ``token = token = ...``
???

File: c2corg_api/security/roles.py [hunk 2]
```diff
@@ -67,6 +71,22 @@ def try_login(user, password, request):
...
+        return add_or_retrieve_token(token, exp, user.id)
+
+    return None
+
+
+def renew_token(user, request):
+    old_token = extract_token(request)
+    valid = validate_token(old_token)
+
+    if valid:
```
Comment (asaunier) [inline c2corg_api/security/roles.py]: Why not
```
if validate_token(old_token):
```
instead?

File: c2corg_api/security/roles.py [hunk 3]
```diff
@@ -67,6 +71,22 @@ def try_login(user, password, request):
...
+        policy = request.registry.queryUtility(IAuthenticationPolicy)
+        now = datetime.datetime.utcnow()
+        exp = now + datetime.timedelta(weeks=CONST_EXPIRE_AFTER_DAYS)
+        claims = create_claims(user, exp)
+        token_value = policy.encode_jwt(request, claims=claims)
+        token = add_or_retrieve_token(token_value, exp, user.id)
+        return token
```
Comment (asaunier) [inline c2corg_api/security/roles.py]:
```
return add_or_retrieve_token(token_value, exp, user.id)
```
???

**Comment by asaunier** [comment]
A few comments else no big objection. Travis fails with some permission problem!?

File: c2corg_api/security/roles.py (hunk 1 above)
Comment (gberaudo) [inline c2corg_api/security/roles.py]: Thanks, strange that the linter does not catch this.

**Comment by gberaudo** [comment]
Thanks for the review @asaunier.
I removed the local variables and reworked the test.

**Comment by asaunier** [comment]
OK for me

**Comment by gberaudo** [comment]
Is it fine with you @tsauerwein?

**Comment by tsauerwein** [comment]
When renewing a token I would have expected that the old token is removed from the database to avoid that there are unused but still valid token in the database.

But could this be a problem? For example from the UI you make a request to renew the token. Before you get the response, you make another authenticated request still with the old token. But in the mean time the old token was removed, so the request would fail.

So, it seems to be good to keep the old tokens. We just have to make sure that we don't renew to often.

**Comment by tsauerwein** [comment]
And yes, this looks good to me!

**Comment by gberaudo** [comment]
Fixed the computed expiration and added a test.
@tsauerwein, I agree with your analyze.

**Comment by gberaudo** [comment]
Thanks for the reviews @asaunier and @tsauerwein.

**Pull request merged**